OCPEDGE-1458: Addressing post-architecture review feedback

jaypoulz opened the pull request:
Updates include:
- Clarifications about the fencing network needing to be seperate
- Calling out unsuitable work loads (e.g. safety-critical)
- Explaining why upgrades only work when both nodes are healthy
- Comparing TNF to active-passive SNO
- Added test for verifying certification rotation with an unhealthy node
- Updated baremetal usage to match https://github.com/metal3-io/metal3-docs/blob/master/design/bare-metal-style-guide.md

Also unified the document with the 200 character word wrap (second commit).

jaypoulz committed:
OCPEDGE-1458: [TNF] Addressing post-architecture review feedback

Updates include:
- Clarifications about the fencing network needing to be separate
- Calling out unsuitable work loads (e.g. safety-critical)
- Explaining why upgrades only work when both nodes are healthy
- Comparing TNF to active-passive SNO
- Added test for verifying certification rotation with an unhealthy node
- Updated baremetal usage to match https://github.com/metal3-io/metal3-docs/blob/master/design/bare-metal-style-guide.md

clobrano commented on enhancements/two-node-fencing/tnf.md:
```diff
@@ -60,20 +60,28 @@ tracking-link:
 
-Leverage traditional high-availability concepts and technologies to provide a container management solution that has a minimal footprint but remains resilient to single node-level failures suitable for customers with numerous geographically dispersed locations.
+Leverage traditional high-availability concepts and technologies to provide a container management solution that has a minimal footprint but remains resilient to single
+node-level failures suitable for customers with numerous geographically dispersed locations.
```
Are these forced `newline` intended? There is a number of them, but it doesn't seem that all the lines respect the same "text with", I wonder if it is the result of some automatic editor formatting

jaypoulz replied:
I've added a new commit that forces wrap at 200 characters. This is just to keep everything aligned with the widest table. Hopefully this should help keep things more standardized.

clobrano commented on enhancements/two-node-fencing/tnf.md:
```diff
@@ -199,8 +209,9 @@ For a two-node cluster to be successful, we need to ensure the following:
-Such operations will only proceed when both peers are online and healthy.
+As per a standard 3-node control-plane, OpenShift upgrades and `MachineConfig` changes that would trigger a node reboot cannot proceed if the aformentioned reboot
+would go over the `maxUnavailable` allowance specified in the machine config pool, which defaults to 1. For a two-node control plane, `maxUnavailable` should only ever
+be set to 1 to help ensure that these events only proceed when both peers are online and healthy, or 0 in the case where the adminsitrator wishes to temporarily disable these events.
```
typo: adminsitrator -> administrator

clobrano commented on enhancements/two-node-fencing/tnf.md:
```diff
@@ -199,8 +209,9 @@ For a two-node cluster to be successful, we need to ensure the following:
 
-As per a standard 3-node control-plane, OpenShift upgrades and `MachineConfig` changes can not be applied when the cluster is in a degraded state.
-Such operations will only proceed when both peers are online and healthy.
+As per a standard 3-node control-plane, OpenShift upgrades and `MachineConfig` changes that would trigger a node reboot cannot proceed if the aformentioned reboot
```
typo (I think). I think it's aforEmentioned

clobrano commented on enhancements/two-node-fencing/tnf.md:
```diff
@@ -426,6 +438,27 @@ Tools for extracting support information (must-gather tarballs) will be updated
+disruption experienced between the nodes on the main network can still proceed with a recovery operation with an available fencing network.
+
+Addressing the matter of ensuring that fencing is configured properly is more challenging. While the installer should protect from some kinds of invalid configuration
+(e.g. an incorrect BMC password), this is not a guarantee that the configuration is actually valid. For example, a user could accidently enter the BMC credentials for
```
Typo: accidently -> accidentally

clobrano commented on enhancements/two-node-fencing/tnf.md:
```diff
@@ -751,9 +784,26 @@ a close eye on when evaluating upgrade jobs for potential race conditions.
+Single-Node OpenShift instances and setting them up as an active node with a passive backup.
+
+The advantage of this this kind of set up is that failures individual nodes in the cluster are reliably deterministic. A failure of your active node just passes traffic over
+to your passive node, and a failure of your passive node doesn't interupt the cluster function at all. Since both nodes cannot be active at the same time, there is no risk
```
typo: interupt -> interrupt

clobrano commented on enhancements/two-node-fencing/tnf.md:
```diff
@@ -751,9 +784,26 @@ a close eye on when evaluating upgrade jobs for potential race conditions.
+This means that the cluster administrator needs to actively synchronize both nodes so that either of them could become the active node at any given time. For some
+applications, this results in more overhead on each node since both need to be able to run fully independently. More importantly, the end user cannot utilize the full
+capacity of both nodes at the same time. Especially at scale, this brings us to the motivating use cases driving Two Node OpenShift with Fencing - applications that can
+trade-off some of the determinism provided by an Active-Passive Single Node in exchange for a simplified adminsitrative experience for managing both nodes while being able
```
typo: adminsitrative -> administrative

clobrano reviewed (changes requested):
I noticed few typos, otherwise lgtm

slintes commented on enhancements/two-node-fencing/tnf.md:
```diff
@@ -339,109 +356,138 @@ sshKey: ''
+of running a 3-node hyperconverged cluster - especially in regard to auto-recovering from the failure of either node. The key to delivering on these goals is fencing.
+
+In order be able to leverage fencing, two conditions must be met:
+1. The fencing network must be available
```
I'm wondering if we need more details about how to add a fencing network, or if network people know what to do?

jaypoulz replied:
We do - but I expect those details to be included via @mshitrit's work on https://issues.redhat.com/browse/OCPEDGE-1461

jaypoulz replied:
This commit just focused on rewrapping the comments that were already there.

slintes replied:
ok

slintes commented on enhancements/two-node-fencing/tnf.md:
```diff
@@ -561,54 +610,49 @@ This proposal is an alternative architecture to Single-node and MicroShift, so i
+   is the cluster-etcd-operator. Ideally, we could have it be responsible for kicking off the initialization of pacemaker, since the core of a successful TNF setup is to ensure etcd ownership is
+   transitioned to a healthy RHEL-HA deployment. While it is a little unorthodox for a core operator to initialize an external component, that component is tightly coupled with the health of etcd to
+   begin with and they benefit from being deployed and tested together.  Additionally, most cases that would result in pacemaker failing to initialize would result in CEO being degraded as well. One
+   concern raised for this approach is that we may introduce a greater security risk since CEO permissions need to be elevated so that a container can run as root to initialize pacemaker. The other
```
> CEO permissions need to be elevated so that a container can run as root

I'm still investigating this, it's more complicated than root or non-root, there are multiple security related settings in a pod / container definition

jaypoulz replied:
I expect this to be updated to address these details as part of https://issues.redhat.com/browse/OCPEDGE-1461. My goal for this section was to mostly just get something in the section that roughly maps to our current understanding.

jaypoulz replied:
I didn't actually change this section. I just added a new commit that has my editor wrap the text at 200 characters so we can avoid manually wrapping and unwrapping sections.

200 characters was chosen because that's the shortest width I could make the widest table.

slintes commented on enhancements/two-node-fencing/tnf.md:
```diff
@@ -175,72 +181,84 @@ Three aspects of cluster creation need to happen for a vanilla two-node cluster
-Once the cluster has two members, the etcd daemon will be removed from the static pod definition and recreated as a resource controlled by RHEL-HA.
-At this point, the cluster-etcd-operator (CEO) will be made aware of this change so that some membership management functionality that is now handled by RHEL-HA can be disabled.
-This will be achieved by having the same entity that drives the configuration of RHEL-HA use the OpenShift API to update a field in the CEO's `ConfigMap` - which can only succeed if the control-plane is healthy.
+Initially, the creation of an etcd cluster will be driven in the same way as other platforms. Once the cluster has two members, the etcd daemon will be removed from the static pod definition and
```
I need to review this paragraph again when I'm more familiar with CEO...

jaypoulz replied:
As with some of the others, I think it would be OK to update this in a separate PR. The goal from my side was to make sure there was a section where folks could write and adjust as part of their own investigations.

beekhof commented on enhancements/two-node-fencing/tnf.md:
```diff
@@ -751,9 +785,26 @@ a close eye on when evaluating upgrade jobs for potential race conditions.
 ## Alternatives
 
-* MicroShift was considered as an alternative but it was ruled out because it does not support multi-node and has a very different experience than OpenShift which does not match the TNF initiative which is on getting the OpenShift experience on two nodes
+#### Active-Passive Single-Node OpenShift
```
My major concerns with this section is that we're glossing over

1. What entity can ensure exactly one active cluster is chosen
2. What entity will route traffic to that cluster
3. How data is synchronised between the clusters, and in what direction
4. The need for bespoke configurations of (and possibly functional modifications to) Red Hat and partner workloads that expect to run multiple instances within a single configuration domain/cluster.

which make this alternative seem more realistic than it is.  
We should make it clear that this option faces all the same challenges as TNF, needs to solve them in the general case, plus introduces additional ones.

jaypoulz replied:
I did my best to address this feedback in the last round. Let's review tonight if possible

jaypoulz committed:
OCPBUGS-1468: [TNF] Added a section to address running with a failed node

- Topology transitions may be discussed later.
- Running TNF with a failed node is not the same as Single Node OpenShift

jaypoulz committed:
[TNF] Unified document around a 200 column wrap to eliminate manual wrapping.

beekhof commented on enhancements/two-node-fencing/tnf.md:
```diff
@@ -687,76 +781,94 @@ Additionally, it would be good to have workload-specific testing once those are
+This option requires the user to run external software like keepalived in front of the instances to ensure that only one instance is active and receiving traffic at a time.
+
+The advantage of this this kind of set up is that failures of each instance is reliably deterministic. A failure of your active instance just passes traffic over to your passive instance, and a
+failure of your passive instance doesn't interrupt the active instance at all. Since both instances cannot be active at the same time, there is no risk of a split-brain situation. Additionally, each
```
s/active/the active instance/ to make it clear that we don't mean it in the "running" sense.

beekhof commented on enhancements/two-node-fencing/tnf.md:
```diff
@@ -687,76 +781,94 @@ Additionally, it would be good to have workload-specific testing once those are
+
+The advantage of this this kind of set up is that failures of each instance is reliably deterministic. A failure of your active instance just passes traffic over to your passive instance, and a
+failure of your passive instance doesn't interrupt the active instance at all. Since both instances cannot be active at the same time, there is no risk of a split-brain situation. Additionally, each
+Single Node OpenShift instance is a cluster by itself - meaning it does need the other instance to be operational for it to proceed with upgrade or configuration operations that require a node reboot.
```
I think you forgot a negation

```suggestion
Single Node OpenShift instance is a cluster by itself - meaning it does not need the other instance to be operational for it to proceed with upgrade or configuration operations that require a node reboot.
```

beekhof commented on enhancements/two-node-fencing/tnf.md:
```diff
@@ -687,76 +781,94 @@ Additionally, it would be good to have workload-specific testing once those are
+#### 2 SNO + KCP:
+[KCP](https://github.com/kcp-dev/kcp/) allows you to manage multiple clusters from a single control-plane, reducing the complexity of managing each cluster independently. With kcp, you can manage the
+two single-node clusters, each single-node OpenShift cluster can continue to operate independently even if the central kcp management plane becomes unavailable. The main advantage of this approach is
+that it doesn’t require inventing a new Openshift flavor and we don’t need to create a new installation flow to accommodate it.
```
How would workloads be affected?
For example, if I'm a deployment with replica=2, would it be possible to arrange for 1 replica to be on each SNO?

jaypoulz replied:
That's not really possible. I tried to address this point in a sentence about workloads with multiple instances, but I'll be even more specific.

mshitrit commented on enhancements/two-node-fencing/tnf.md:
```diff
@@ -518,9 +585,39 @@ This section provides specific steps for how two-node clusters would handle inte
+#### Regarding Compute Nodes
+
+As currently envisioned, it doesn't make sense for a customer to install a new TNF cluster with compute nodes. If the cluster has compute capacity for 3 nodes, it should always be installed as a
+highly available (3 control-plane) cluster. This beomes more complicated for customers that discover after installing a fleet of TNF clusters that they need to add compute compacity to meet a rise in
```
Typo: beomes instead of becomes

mshitrit commented on enhancements/two-node-fencing/tnf.md:
```diff
@@ -565,6 +565,14 @@ possible, but it is not sufficient for the operation of a healthy cluster longte
 
+#### Regarding Compute Nodes
+
+As currently envisioned, it doesn't make sense for a customer to install a new TNF cluster with compute nodes. If the cluster has compute capacity for 3 nodes, it should always be installed as a highly available (3 control-plane) cluster. This beomes more complicated for customers that discover after installing a fleet of TNF clusters that they need to add compute compacity to meet a rise in dema ...
```
I think you meant `compute capacity` instead of `compute compacity`

mshitrit commented on enhancements/two-node-fencing/tnf.md:
```diff
@@ -565,6 +565,14 @@ possible, but it is not sufficient for the operation of a healthy cluster longte
 
 Because of the flexibility it can offer users to adjust their cluster capacity according to their needs, transitioning between control-plane topologies may be introduced for evaluation in a future enhancement.
 
+#### Regarding Compute Nodes
```
Nit: I think it might be worth to clarify that this discussion is about `adding` additional compute nodes, which might be trivial but isn't stated

jaypoulz replied:
I changed the title, but really it covers installing with compute nodes and adding them later.

mshitrit commented on enhancements/two-node-fencing/tnf.md:
```diff
@@ -565,6 +565,14 @@ possible, but it is not sufficient for the operation of a healthy cluster longte
+
+The alternative that we expect to be asked for is to support compute nodes in TNF. For this reason, we'd like to avoid investing heavily in preventing compute nodes from joining the cluster - at install or runtime. It's worth noting that no variant of OpenShift currently actively prevents compute nodes from being added to the cluster.
+
+The only boundary we'd set to declare compute nodes as unsupported in documentation.
```
My apologizes if my question was already discussed :-)
But this statement seems a bit counter intuitive to general rule of thumb in which we try to make sure that we prevent users from taking unsupported actions.
I'm wondering as to the reason for this exception :thinking:

jaypoulz replied:
The point of this section is that we don't intend to build out special validation just to tear it down in a future release. At a certain point, introducing tech debt outweighs the benefits of focusing the scope of the user experience.
My other argument is that the installer generally allows anybody to install anything. That's why we didn't need to change anything to install with a 2-node control plane.

mshitrit commented on enhancements/two-node-fencing/tnf.md:
```diff
@@ -551,6 +551,20 @@ This section provides specific steps for how two-node clusters would handle inte
+An interesting aspect of TNF is that should a node fail and remain in a failed state, the cluster recovery operation will allow the survivor to restart etcd as a cluster-of-one and resume normal
+operations. In this state, we have an operation running cluster with a single control plane node. So, what is the difference between this state and running Single Node OpenShift? There are three key aspects:
+
+1. Operators that deploy to multiple nodes will become degraded.
+2. Operations that would violate pod-disruption budgets will not work.
+3. Lifecycle operations that would violate the `MaxUnavailable` setting of the control-plane MachineSet cannot proceed. This includes MCO node reboots and cluster upgrades.
```
Do we explain how will this be done  ?

jaypoulz replied:
This happens automatically because the logic related to the Machine config's MaxUnavailable default is set to 1. Basically MCO will not signal for the machine to reboot if it would violate this. I don't know that there is much to explain. I could link it to the appropriate documentation - that might be helpful.

jaypoulz committed:
[TNF] Added a note explaining stance on adding compute nodes
